TMC2208 and TMC2209 support (includes improvements)

V1pr opened the pull request:
Hi,

I've continued tiagojbalmeida's #1329  and LastDragon-ru's #1377 work. Added TMC2209 support beside TMC2208 with coolconf and stallguard support. Merged beeverycreatie's work (PR #1315 ) as well, to reflect MotorDriverControl's structure change.

- Since the MotorControl module was getting messy for me with adding a new driver I've reworked and added a virtual parent class (StepperDrv) between the MotorControl class and the actual driver class. To don't break compatibility I've modified the function names in the subclasses (mostly in TMC26X). Naming conventions as in DRV8711.
- renamed the module to TMC220X to reflect the compatibility a bit more
- added example file in config snipplets
- driver works on my MK S board with TMC2209 (also in TMC2208 mode), UART ok, can set the current/microsteps from smoothie etc
- TMC21X driver's typo have been corrected - chopper_mode was set from resistor in init

tiagojbalmeida committed:
Basic UART communication with TMC2208 stepper motor driver

tiagojbalmeida committed:
Define and document all registers

tiagojbalmeida committed:
Move SoftSerial and Modbus to libs folder

tiagojbalmeida committed:
Use Software UART libs to communicate with TMC2208 driver

tiagojbalmeida committed:
Import functions to configure driver via UART

tiagojbalmeida committed:
Add StealthChop and StealthChop threshold functions

tiagojbalmeida committed:
Add TMC2208 config file

tiagojbalmeida committed:
Update velocity threshold

tiagojbalmeida committed:
Move print information about successfull configured motor drivers

tiagojbalmeida committed:
Add TMC2208 stepper motor driver support

tiagojbalmeida committed:
Move Modbus lib back to spindle module

tiagojbalmeida committed:
Put SPI and UART pins in an union to save memory

tiagojbalmeida committed:
Delete UART object after initial driver configuration

tiagojbalmeida committed:
Put SPI and UART objects inside union to save memory

tiagojbalmeida committed:
Fix SPI/UART union structure

tiagojbalmeida committed:
Remove driver setup condition

tiagojbalmeida committed:
Disable stealthChop threshold as default

tiagojbalmeida committed:
Set general configuration register

tiagojbalmeida committed:
set slaveconf register

tiagojbalmeida committed:
set HOLD parameters

tiagojbalmeida committed:
Add tpowerdown register

tiagojbalmeida committed:
Fix mode parameter name

LastDragon-ru committed:
Added `motor_driver_control.timeout`.

LastDragon-ru committed:
Added `motor_driver_control.driver.tpwmthrs`.

LastDragon-ru committed:
Fixed `motor_driver_control.driver.microsteps` description.

LastDragon-ru committed:
Write only mode support.

LastDragon-ru committed:
Config description and small fixes

LastDragon-ru committed:
Optimized chip configuration (reduced number of writes)

LastDragon-ru committed:
More settings:
* `motor_driver_control.enable`
* `motor_driver_control.driver.tpwmthrs` renamed to `motor_driver_control.driver.stealthchop_tpwmthrs`
* `motor_driver_control.driver.spreadcycle_toff`
* `motor_driver_control.driver.spreadcycle_tbl`
* `motor_driver_control.driver.spreadcycle_hstrt`
* `motor_driver_control.driver.spreadcycle_hend`

LastDragon-ru committed:
Forced `setStepInterpolation(1)`

LastDragon-ru committed:
Stealthchop settings support.

LastDragon-ru committed:
Removed tx/rx `port_number` shifting.

V1pr committed:
created parent class for drivers, making the main code cleaner. Tuned TMC22X class, but not (yet) tested. In the other two classes just some function naming changes due to the parent class.

V1pr committed:
TMC22X still with debug messages (but fewer), compiles ok.
Fixed fix actuator in dump_status; dump_status should work better now with online data from CHIP.

V1pr committed:
changed current for set_current from uint32_t to uint16_t (65535 mA should be enough, but we save some memory)

V1pr committed:
fixed missing type conversions. Compiles ok.

V1pr committed:
M122 debugging (for TMC22X now)

V1pr committed:
debug msg fine tune

V1pr committed:
added/finished TMC2209 support; renamed helper namespace to StepstickParameters

V1pr committed:
renamed TMC22X to TMC220X to be more informative

V1pr committed:
added TMC2209 registers descriptors

V1pr committed:
more coolconf settings, set_options handles coolconf for TMC220X

V1pr committed:
added credits, typo fix and example tmc2209 config

wolfmanjm commented:
does this completely replace those other two PRs? so I can close them in favor of this one?

V1pr commented:
> does this completely replace those other two PRs? so I can close them in favor of this one?

Yes, includes both (from a functionality perspective).

wolfmanjm commented on src/modules/utils/motordrivercontrol/MotorDriverControl.cpp:
```diff
@@ -54,6 +63,21 @@ MotorDriverControl::~MotorDriverControl()
+
+    if (timeout > 0)
+    {
+        wait(timeout);
```
If wait does what I think it does it is not allowed. it is not allowed to halt processing for any reason at anytime.

V1pr replied:
It's exaclty as it seems, however it's a heritage from the previous versions; I'm not sure if it's really needed, but since it default's to 0 and used only at init time I don't think it's a big issue to have it here. Anyway, if you say so, I'l remove it.

wolfmanjm replied:
yes it needs to be removed. wait() is NOT a valid call that is allowed anywhere in smoothie. I must have missed it in the other requests.

V1pr replied:
Removed as requested.

wolfmanjm commented:
I like the refactor to clean up the motorcontrol class.

wolfmanjm commented:
yes it needs to be removed. wait() is NOT a valid call that is allowed anywhere in smoothie. I must have missed it in the other requests.

V1pr commented:
> I like the refactor to clean up the motorcontrol class.

Maybe it's possible to merge StepperDrv class into MotorCotorcontrol, but neither am I that good at C++, nor is this PR good for such a great change.

V1pr committed:
removed timeout for driver init

wolfmanjm commented on src/modules/utils/motordrivercontrol/MotorDriverControl.cpp:
```diff
@@ -54,6 +63,21 @@ MotorDriverControl::~MotorDriverControl()
 void MotorDriverControl::on_module_loaded()
 {
+    // Enabled?
+    if (!THEKERNEL->config->value(motor_driver_control_checksum, enable_checksum )->by_default(1)->as_bool()) {
```
by_default needs to be false here, also not sure how that slipped in.

wolfmanjm replied:
It is actually superfluous not sure how it got in there as my original code did not have this at all. But it can't do any harm I suppose. I does make old configs not backward compatible though, so it should not have been added at all.

V1pr replied:
should I remove these lines? The loop below this one checks anyway if a module is enabled or not.

wolfmanjm replied:
yes you can it serves no purpose.

V1pr committed:
motorcontrol default OFF

V1pr commented:
Note: there might be an error in TMC26X, but did not want to change anything 'real' in the other drivers. I thing coolconf register SEDN and SEUP are mixed up (decrement is starting at bit 13, while increment is at 5 according to the docs. I don't know who was the author of the original driver, but IMHO this should be checked separatelly (see line 630 in TMC26X.cpp)
`
    cool_step_register_value = ((unsigned long)lower_SG_threshold) | (((unsigned long)SG_hysteresis) << 8) | (((unsigned long)current_decrement_step_size) << 5)
                               | (((unsigned long)current_increment_step_size) << 13) | (((unsigned long)lower_current_limit) << 15)
                               //and of course we have to include the signature of the register
                               | COOL_STEP_REGISTER;
`

mvadu commented:
@V1pr does this PR support SPI TMC drivers too?

V1pr commented:
> @V1pr does this PR support SPI TMC drivers too?

Yes, and no :) This PR is for TMC220X which is an UART driver. However originally there is/was the TMC26X driver for TMC2660, which is an SPI chip. To give you some secret info: once this PR is accepted I'll start working on support for TMC5161, which is SPI.

V1pr committed:
removed unneded motorcontrol.enable check

vbolshakov commented:
@V1pr Please provide your connection diagram. I try to test your work, but unable to change even microsteps setting.

vbolshakov commented:
Able to partially test functionality - change TX and RX from I2C pins, that are pull-up to 3.3 on Smoothieboard.
Current control works - using M906X20 stepper start loosing steps.
Microstepping not work. Not from config, not from M909.

V1pr commented:
@vbolshakov that's interesting. If current control works, than the other should as well. What does M122 os M911 say? Do the registers change? 'M909 X256' has no effect? I tested my 2209 with a nema23, which only started to move if I raised the current to 2A and set the microsteps to 256, otherwise there was not enogh power to move the carrige.

vbolshakov commented:
M122 and M911 gives unreliable output - parameters changed  from read to read.

vbolshakov commented:
This is my config:
motor_driver_control.enable                     true     
motor_driver_control.alpha.enable               true     
motor_driver_control.alpha.axis                 X        
motor_driver_control.alpha.chip                 TMC2208  
motor_driver_control.alpha.sw_uart_tx_pin       1.29     
motor_driver_control.alpha.sw_uart_rx_pin       2.1      
motor_driver_control.alpha.sw_uart_baudrate    19200     
motor_driver_control.alpha.current             200       
#motor_driver_control.alpha.sense_resistor      50       
motor_driver_control.alpha.microsteps          128       
#motor_driver_control.alpha.chopper_mode        0        
This is boot loog:
etting GCONF to 000001C5 for axis X
GCONF status: 33EFBEAD (871349933) [CRC: 1]
DRV status: 33EFBEAD (871349933) [CRC: 1]
MotorDriverControl INFO: configured motor X (1): as TMC2208, tx: 011D, rx: 0201

M911 output:
M911
Motor 1 (X)...
designator X [0], Chip type TMC2208
Microstep position phase A: -117440784
Current setting: 160mA
Microsteps: 1/128
X - ERROR: Overtemperature Shutdown!
X - ERROR: SHORT to ground on channel B!
X - ERROR: Channel A seems to be unconnected!
X - ERROR: Channel B seems to be unconnected!
Register dump [L=from uC memory]:
 gconf register: 000001C5 (453) C7FFFFFC
 slaveconf register: 00000000 (0)
 ihold_irun register: 00000100 (256)
 tpowerdown register: 00000014 (20)
 tpwmthrs register: 00000000 (0)
 chopconf register: 110301B5 (285409717)
 pwmconf register: C10D0024 (-1056112604)
S#-742342657,D1,C160,M-1,Ao,Bo,Te,EN1,Co5,Cb32,

Or:
Motor 1 (X)...
designator X [0], Chip type TMC2208
Microstep position phase A: 268406776
Current setting: 160mA
Microsteps: 1/64
X - WARNING: Overtemperature Prewarning!
Register dump [L=from uC memory]:
 gconf register: 000001C5 (453) D2C0FE9F
 slaveconf register: 00000000 (0)
 ihold_irun register: 00000100 (256)
 tpowerdown register: 00000014 (20)
 tpwmthrs register: 00000000 (0)
 chopconf register: 110301B5 (285409717)
 pwmconf register: C10D0024 (-1056112604)
S#-1769472064,D1,C160,M2,Ag,Bo,Te,EN1,Co5,Cb32,

M122X
GCONF status for X: D6FCFFFF (3606904831)
DRV status for X: 0FFF80C0 (268402880)
ok
M122X
GCONF status for X: CDC08FFF (3451949055)
DRV status for X: 829FC0FE (2191507710)
ok
M122X
GCONF status for X: 85FFFF8F (2248146831)
DRV status for X: 4FFFFEE1 (1342176993)

vbolshakov commented:
@V1pr what board do you use? I will try SKR 1.3, but need to remove it from my 3D printer...

V1pr commented:
@vbolshakov I've tried it with an M kS s gen. Bootup looks ok, the 'CRC: 1' indicates, that the read value is ok (based on the crc got/calculated). M122 output can change, anything is good if it's not 00000000 or ffffffff (reg. values change). Did you change the microsteps between the two 911 output?

vbolshakov commented:
@V1pr no, it is just several sequential M911...

vbolshakov commented:
On SKR 1,3 using RX@2.12 and TX@2.13 (with 1.6k resistor) i'm able to set microstepping. But reading from driver produce garbage.
Main difference is CPU working at 100MHz.

@wolfmanjm , is it possible to use hardware UART1?

V1pr commented:
@vbolshakov 1.6k is too much and that can cause the troubles. In the docs it says 1k, I suggest you change it. Moreover, there is no stepper/power-supply beside the rx/tx line? Than can cause problems as well, if you don't shield the cable.

Theorically it's possible to use the HW UART, but that I won't do (and will need some ode change). What would be feasible is to use an I2C GPIO expander or an I2C serial expander.

vbolshakov commented:
USB UART work well with 1.6K and even without it. I also tried recommended value and even connecting directly - reading not work...
Cables not shielded, it it just 5 cm wires with dupont connector hang over board. 

About I2C - as i remember there was problems with i2c timeout on LCP1769 so Smoothie not work with LED displays over i2c.

V1pr committed:
fixed typo for SEIMIN

wolfmanjm commented:
As this changes th ebasic structure of the motordriver, and does not include trhe TMC21X patches it would make that PR unbuildable. So either this needs to include those drivers or this will need to be a separate fork.

V1pr committed:
added TMC21X support from beeverycreative fork (modified function names to reflect new MotorDriverControl structure + minor fixes for init and dump_status)

V1pr committed:
added missing config snippet

V1pr commented:
@wolfmanjm TMC21X included now. Fixed the typo mentioned in #1315 ; removed unneeded 'readable' flag from dump_status; added dynamic actuator handling in dump_status (was fix to X with a TODO flag). PR builds ok, but TMC21X will need testing.

V1pr commented:
TMC21X: while checking out the docs I've found, that someone might have failed to read the docs properly for the chip. Thus, the read access for the TMC2130 will not work properly... from the docs: "A read access request datagram uses dummy write data. Read data is transferred back to the master with the subsequent read or write access. Hence, reading multiple registers can be done in a pipelined fashion."
Writing the TMC2130 should not be an issue IMHO, so the 2130 will work in a write only manner with this code. I'll fix it when I have some spare time.

CBJamo commented:
The drivers/Tmc220x folder should be capitalized to match the other drivers. The lowercase letters will cause a build fail on linux.

wolfmanjm commented on src/modules/utils/motordrivercontrol/drivers/Tmc220x/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+/*
```
The directory name needs to be uppercase.

V1pr replied:
Shu :( Fixed.

Commit:
dirname fix

V1pr commented:
@CBJamo sorry, fixed now. Please try again.

mvadu commented:
> > @V1pr does this PR support SPI TMC drivers too?
> 
> Yes, and no :) This PR is for TMC220X which is an UART driver. However originally there is/was the TMC26X driver for TMC2660, which is an SPI chip. To give you some secret info: once this PR is accepted I'll start working on support for TMC5161, which is SPI.

@V1pr in #1315 comments you seems to indicate you have added the TMC2130 into this PR. Is that the case? I have a board with 2130, and can test it if that is the case.

V1pr commented:
> > > @V1pr does this PR support SPI TMC drivers too?
> > 
> > 
> > Yes, and no :) This PR is for TMC220X which is an UART driver. However originally there is/was the TMC26X driver for TMC2660, which is an SPI chip. To give you some secret info: once this PR is accepted I'll start working on support for TMC5161, which is SPI.
> 
> @V1pr in #1315 comments you seems to indicate you have added the TMC2130 into this PR. Is that the case? I have a board with 2130, and can test it if that is the case.

@mvadu: yes, indeed. However, I did not test the 2130 driver just made the necessary modifications to be compatible with the new motordrivercontrol/stepperdrv - everything else is intact, just as in PR #1315

dzhang97 commented:
@V1pr would you be able to add support for the tmc5160? besides the encoder support, it functions very similarly to the tmc2130.

V1pr commented:
> @V1pr would you be able to add support for the tmc5160? besides the encoder support, it functions very similarly to the tmc2130.

@dzhang97 : not yet, did not have time for it. However the 2130 and the 5160 have a lot of things in common, there are some differences; I did not want to create a totally new driver with a lot of copy-paste from the 2130, but extend it and optimize it - and this takes more time than I thought at first.

mvadu reviewed (changes requested):
I tried to pull this into my local repo. I am still getting the folder name as TMC220x, which causes the MotorDriverControl.cpp to fail. I think you need to have dirname as TMC220X with capital X.

Commit:
dirname fix 2

V1pr commented:
> I tried to pull this into my local repo. I am still getting the folder name as TMC220x, which causes the MotorDriverControl.cpp to fail. I think you need to have dirname as TMC220X with capital X.

Darn, I've overlooked somehow last time. Changed.

@mvadu: thanks for pointing this out

mvadu commented:
> > I tried to pull this into my local repo. I am still getting the folder name as TMC220x, which causes the MotorDriverControl.cpp to fail. I think you need to have dirname as TMC220X with capital X.
> 
> Darn, I've overlooked somehow last time. Changed.
> 
> @mvadu: thanks for pointing this out

You are welcome.. somehow your [latest commit](https://github.com/Smoothieware/Smoothieware/pull/1447/commits/c449b8b64aceb2840d2a5c93e98d699609c41de4) makes it look like the cpp and h are added new. For git to detect the directory rename, you need to
1. Rename the folder
2. do a `git status`, it will detect as a new untracked folder, and two files as deleted
3. stage the deleted files ( e.g `git rm src/modules/utils/motordrivercontrol/drivers/TMC220x/TMC220X.cpp`)
4. stage the new folder (e.g. `git add src/modules/utils/motordrivercontrol/drivers/TMC220X`)
5. commit.

This should detect as a folder rename. [see this in my repo](https://github.com/mvadu/Smoothieware/commit/7b4e35d15cd9a2413143dbc4e86ccb324979514d), which shows only the folder name as delta 
```
src/modules/utils/motordrivercontrol/drivers/{TMC220x → TMC220X}/TMC220X.cpp
src/modules/utils/motordrivercontrol/drivers/{TMC220x → TMC220X}/TMC220X.h
```

Again, sorry to nitpick, I did the same thing you did, and had to learn the correct way to do it.

Commit:
dirname fix 3

mvadu commented:
@V1pr I am trying to test this on my extruder motor, with below config. I realized that it does not work, then looking at the code I saw `config_module` actually checks in ([MotorDriverControl.cpp#L96](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L96), and only XYZABC are allowed. Should it also allow E for extruder?

```
## Motor driver configuration
motor_driver_control.driver.enable                 true            # Enable motor driver module
motor_driver_control.driver.axis                   E               # Axis designator on which the motor driver is set 
motor_driver_control.driver.chip                   TMC2130         # Motor driver chip type
motor_driver_control.driver.spi_channel            1               # Set corresponding pins of the selected spi channel
motor_driver_control.driver.spi_cs_pin             0.0             # Set chip select pin
motor_driver_control.driver.current               1000            # Motor currents in mA
motor_driver_control.driver.sense_resistor        110             # Sensing resistor value as a reference in milliohms
motor_driver_control.driver.microsteps            16              # Set microstepping value per step pulse. It can be 1, 2, 4, 8, 16, 32, 64, 128 or 256 microsteps
motor_driver_control.driver.chopper_mode          1               # 0 = stealthChop ; 1 = spreadCycle ; 2 = traditional constant off-time
```

mvadu committed:
initialize connection_method and max current for TMC2130

mvadu commented:
@V1pr  current TMC21X driver does not initialize connection method. So it fails to load the config, and errors out at [MotorDriverControl.cpp#L204](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L204) (e.g. `MotorDriverControl A ERROR: Unsupported connection method! Only SPI and UART supported.` ). 

I have added that code and sent a PR against your branch (https://github.com/V1pr/Smoothieware/pull/2). Please accept it and I will retest.

V1pr committed:
Merge pull request #2 from mvadu/fix_2130

Fix 2130 initialization

V1pr commented:
> > > I tried to pull this into my local repo. I am still getting the folder name as TMC220x, which causes the MotorDriverControl.cpp to fail. I think you need to have dirname as TMC220X with capital X.
> > 
> > 
> > Darn, I've overlooked somehow last time. Changed.
> > @mvadu: thanks for pointing this out
> 
> You are welcome.. somehow your [latest commit](https://github.com/Smoothieware/Smoothieware/pull/1447/commits/c449b8b64aceb2840d2a5c93e98d699609c41de4) makes it look like the cpp and h are added new. For git to detect the directory rename, you need to
> 
> 1. Rename the folder
> 2. do a `git status`, it will detect as a new untracked folder, and two files as deleted
> 3. stage the deleted files ( e.g `git rm src/modules/utils/motordrivercontrol/drivers/TMC220x/TMC220X.cpp`)
> 4. stage the new folder (e.g. `git add src/modules/utils/motordrivercontrol/drivers/TMC220X`)
> 5. commit.
> 
> This should detect as a folder rename. [see this in my repo](https://github.com/mvadu/Smoothieware/commit/7b4e35d15cd9a2413143dbc4e86ccb324979514d), which shows only the folder name as delta
> 
> ```
> src/modules/utils/motordrivercontrol/drivers/{TMC220x → TMC220X}/TMC220X.cpp
> src/modules/utils/motordrivercontrol/drivers/{TMC220x → TMC220X}/TMC220X.h
> ```
> 
> Again, sorry to nitpick, I did the same thing you did, and had to learn the correct way to do it.

NP at all, I'm not a git expert and always keen on learning. I thought I had done the same as before, when renaming, but apparently not.

V1pr commented:
> @V1pr current TMC21X driver does not initialize connection method. So it fails to load the config, and errors out at [MotorDriverControl.cpp#L204](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L204) (e.g. `MotorDriverControl A ERROR: Unsupported connection method! Only SPI and UART supported.` ).
> 
> I have added that code and sent a PR against your branch ([V1pr#2](https://github.com/V1pr/Smoothieware/pull/2)). Please accept it and I will retest.

I'm really confused here, since I did compile the code, but this is missing. Anyway, I've accepted the PR, thanks.

V1pr commented:
> @V1pr I am trying to test this on my extruder motor, with below config. I realized that it does not work, then looking at the code I saw `config_module` actually checks in ([MotorDriverControl.cpp#L96](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L96), and only XYZABC are allowed. Should it also allow E for extruder?
> 
> ```
> ## Motor driver configuration
> motor_driver_control.driver.enable                 true            # Enable motor driver module
> motor_driver_control.driver.axis                   E               # Axis designator on which the motor driver is set 
> motor_driver_control.driver.chip                   TMC2130         # Motor driver chip type
> motor_driver_control.driver.spi_channel            1               # Set corresponding pins of the selected spi channel
> motor_driver_control.driver.spi_cs_pin             0.0             # Set chip select pin
> motor_driver_control.driver.current               1000            # Motor currents in mA
> motor_driver_control.driver.sense_resistor        110             # Sensing resistor value as a reference in milliohms
> motor_driver_control.driver.microsteps            16              # Set microstepping value per step pulse. It can be 1, 2, 4, 8, 16, 32, 64, 128 or 256 microsteps
> motor_driver_control.driver.chopper_mode          1               # 0 = stealthChop ; 1 = spreadCycle ; 2 = traditional constant off-time
> ```

@mvadu I did not change the original motordrivercontrol's logic. I agree, that this should be changed, but I'm not sure if allowing a single "E" is enough (dual extrusion support? Axis is 'char', so it cannot be E0, E1 etc) @wolfmanjm should I change this and allow E,F,G as an axis? Or what would be the proper naming?

wolfmanjm commented:
> @V1pr I am trying to test this on my extruder motor, with below config. I realized that it does not work, then looking at the code I saw `config_module` actually checks in ([MotorDriverControl.cpp#L96](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L96), and only XYZABC are allowed. Should it also allow E for extruder?
> 
> ```
> ## Motor driver configuration
> motor_driver_control.driver.enable                 true            # Enable motor driver module
> motor_driver_control.driver.axis                   E               # Axis designator on which the motor driver is set 
> motor_driver_control.driver.chip                   TMC2130         # Motor driver chip type
> motor_driver_control.driver.spi_channel            1               # Set corresponding pins of the selected spi channel
> motor_driver_control.driver.spi_cs_pin             0.0             # Set chip select pin
> motor_driver_control.driver.current               1000            # Motor currents in mA
> motor_driver_control.driver.sense_resistor        110             # Sensing resistor value as a reference in milliohms
> motor_driver_control.driver.microsteps            16              # Set microstepping value per step pulse. It can be 1, 2, 4, 8, 16, 32, 64, 128 or 256 microsteps
> motor_driver_control.driver.chopper_mode          1               # 0 = stealthChop ; 1 = spreadCycle ; 2 = traditional constant off-time
> ```

NO! E is NOT an axis, it uses the A axis. E is just a hack designator that we have to support due to reprap misuse of designators. Also when used in a G1 command it has special significance as it may be in mm³ insted of mm and is converted to mm by the volumetric calculation.

mvadu commented:
> > @V1pr current TMC21X driver does not initialize connection method. So it fails to load the config, and errors out at [MotorDriverControl.cpp#L204](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L204) (e.g. `MotorDriverControl A ERROR: Unsupported connection method! Only SPI and UART supported.` ).
> > I have added that code and sent a PR against your branch ([V1pr#2](https://github.com/V1pr/Smoothieware/pull/2)). Please accept it and I will retest.
> 
> I'm really confused here, since I did compile the code, but this is missing. Anyway, I've accepted the PR, thanks.

Thank you @V1pr !! Yeah, without those two lines it did compile, but config load failed. Any way, after that it did read the config (with A as axis), and fails to move the motor. `M911` is complaining all errors, have to check if its my chip or the code. Let me know if you see anything obvious. 

```
M911
Motor 1 (A)...
designator A, Chip type TMC21X
A - ERROR: Overtemperature Shutdown!
A - ERROR: SHORT to ground on channel A!
A - ERROR: SHORT to ground on channel B!
A - ERROR: Channel A seems to be unconnected!
A - ERROR: Channel B seems to be unconnected!
INFO: Stall Guard level reached!
INFO: Motor is standing still.
Microstep position phase A: -1
Stall Guard value: 1023
Current setting: 76mA
Coolstep current: 2462mA
Microsteps: 1/16
Register dump:
 gconf register: 00000004 (4)
 ihold_irun register: 00000000 (0)
 tpowerdown register: 00000000 (0)
 tpwmthrs register: 00000000 (0)
 tcoolthrs register: 00000000 (0)
 thigh register: 00000000 (0)
 chopconf register: 04000000 (67108864)
 coolconf register: 00000000 (0)
 pwmconf register: 00050480 (328832)
 motor_driver_control.xxx.reg 00004,00000,00000,00000,00000,00000,4000000,00000,50480
#s,d1,c76,m16,t0,f0,Ke-,Kl7648,Ku0,Kn0,Ki0,Km0,ao,bo,xe,e0,Cm0,Co2,Cb-17,Cs-3,Ce-6,Cd-2,
```

V1pr commented:
> > > @V1pr current TMC21X driver does not initialize connection method. So it fails to load the config, and errors out at [MotorDriverControl.cpp#L204](https://github.com/Smoothieware/Smoothieware/blob/e8306f674956cd6b18d87db88b1f0b9988a55678/src/modules/utils/motordrivercontrol/MotorDriverControl.cpp#L204) (e.g. `MotorDriverControl A ERROR: Unsupported connection method! Only SPI and UART supported.` ).
> > > I have added that code and sent a PR against your branch ([V1pr#2](https://github.com/V1pr/Smoothieware/pull/2)). Please accept it and I will retest.
> > 
> > 
> > I'm really confused here, since I did compile the code, but this is missing. Anyway, I've accepted the PR, thanks.
> 
> Thank you @V1pr !! Yeah, without those two lines it did compile, but config load failed. Any way, after that it did read the config (with A as axis), and fails to move the motor. `M911` is complaining all errors, have to check if its my chip or the code. Let me know if you see anything obvious.
> 
> ```
> M911
> Motor 1 (A)...
> designator A, Chip type TMC21X
> A - ERROR: Overtemperature Shutdown!
> A - ERROR: SHORT to ground on channel A!
> A - ERROR: SHORT to ground on channel B!
> A - ERROR: Channel A seems to be unconnected!
> A - ERROR: Channel B seems to be unconnected!
> INFO: Stall Guard level reached!
> INFO: Motor is standing still.
> Microstep position phase A: -1
> Stall Guard value: 1023
> Current setting: 76mA
> Coolstep current: 2462mA
> Microsteps: 1/16
> Register dump:
>  gconf register: 00000004 (4)
>  ihold_irun register: 00000000 (0)
>  tpowerdown register: 00000000 (0)
>  tpwmthrs register: 00000000 (0)
>  tcoolthrs register: 00000000 (0)
>  thigh register: 00000000 (0)
>  chopconf register: 04000000 (67108864)
>  coolconf register: 00000000 (0)
>  pwmconf register: 00050480 (328832)
>  motor_driver_control.xxx.reg 00004,00000,00000,00000,00000,00000,4000000,00000,50480
> #s,d1,c76,m16,t0,f0,Ke-,Kl7648,Ku0,Kn0,Ki0,Km0,ao,bo,xe,e0,Cm0,Co2,Cb-17,Cs-3,Ce-6,Cd-2,
> ```

@mvadu  looks like a connectivity issue to me (use the proper SPI channel 0/1 and have the correct EN pin connected as well - careful to use a really free pin). The TMC21X code is just "lifted up" from the original source with minimal changes to reflect the new structure.

mvadu commented:
@V1pr Thank you!! I have a custom smoothieboard, and its schematic had wrong pin showing for CSN, once cross checked its working nicely.

mvadu commented:
@V1pr I am testing with [Bigtreetech sticks ](https://github.com/bigtreetech/BIGTREETECH-TMC2130-V3.0). Before this firmware I was using with jumpers, and I was getting extremely smooth motion as well as dead silent motors. I was running the motos at around 1.2A tuned using the pot.

```
Config / Selected / Meaning
CFG0 = VCC = Tclk=236, Toff=7
CFG1, CFG2 = Open = 16 µsteps in SteachChop, MRES=4, intpol=1, en_PWM_mode=1
CFG3 = Open = Current scale setby sense resistors and scaled by AIN. I_scale_analog=1 
```

I tried this config with your code
```
#alpha
motor_driver_control.alpha.enable                   true       
motor_driver_control.alpha.axis                     X          
motor_driver_control.alpha.chip                     TMC2130    
motor_driver_control.alpha.spi_channel              1          
motor_driver_control.alpha.spi_cs_pin               1.20       
motor_driver_control.alpha.current                  1000       
motor_driver_control.alpha.sense_resistor           110        
motor_driver_control.alpha.microsteps               32         
motor_driver_control.alpha.chopper_mode             0          
```

Even after trying I can get the motors moving but not smooth, there is a low pitch whine. Could you please help me to tune this config to match what I can get with jumpers?

mvadu committed:
Fix set_current signature

mvadu committed:
Fix current calculation

mvadu commented:
> @V1pr I am testing with [Bigtreetech sticks ](https://github.com/bigtreetech/BIGTREETECH-TMC2130-V3.0). Before this firmware I was using with jumpers, and I was getting extremely smooth motion as well as dead silent motors. I was running the motos at around 1.2A tuned using the pot.
> 
> ```
> Config / Selected / Meaning
> CFG0 = VCC = Tclk=236, Toff=7
> CFG1, CFG2 = Open = 16 µsteps in SteachChop, MRES=4, intpol=1, en_PWM_mode=1
> CFG3 = Open = Current scale setby sense resistors and scaled by AIN. I_scale_analog=1 
> ```
> 
> I tried this config with your code
> 
> ```
> #alpha
> motor_driver_control.alpha.enable                   true       
> motor_driver_control.alpha.axis                     X          
> motor_driver_control.alpha.chip                     TMC2130    
> motor_driver_control.alpha.spi_channel              1          
> motor_driver_control.alpha.spi_cs_pin               1.20       
> motor_driver_control.alpha.current                  1000       
> motor_driver_control.alpha.sense_resistor           110        
> motor_driver_control.alpha.microsteps               32         
> motor_driver_control.alpha.chopper_mode             0          
> ```
> 
> Even after trying I can get the motors moving but not smooth, there is a low pitch whine. Could you please help me to tune this config to match what I can get with jumpers?

I found the problem. TMC2130 code had wrong signature for `set_current`, instead of `set_current(uint16_t current)` it had `set_current(unsigned int current)` which made the initialization to skip setting current. So it was running with default 76mA of motor current. 

Another problem I found (which I think is there in 22X driver as well) is wrong formula used for current calculation. Datasheet for TMC2130 has this formula
![image](https://user-images.githubusercontent.com/8618235/86998151-c2b5f980-c17d-11ea-900c-df786e65dba9.png)

But in the code, somehow they missed the `∗ 1/√2` part in [set_current](https://github.com/V1pr/Smoothieware/blob/873a364b946e69d5e9a0607bf82cbf44602ffe98/src/modules/utils/motordrivercontrol/drivers/TMC21X/TMC21X.cpp#L1271) They start with with `I=(CS+1)/32*(Vsense/Rsense)` and solve for CS. I am not good with Math, and I used [wolframalpha](https://www.wolframalpha.com/input/?i=%28%28%28C+%2B1%29%2F32%29+*+%28V%2F%28%28R%2F1000%29+%2B%2820%2F1000%29%29%29+*+1%2F%E2%88%9A2%29%3DX+for+C) and got `CS = (4 sqrt(2) (Rs + 20) I)/(125 Vfs) - 1`. Simplifying for constants, I think we should use `C = (5.65685F * (resistor_value + 20) *mASetting/1000.0F)/(125 * Vfs) - 1`, where Vfs will be 0.32 or 0.18. 

I have pushed a fix to your branch (https://github.com/V1pr/Smoothieware/pull/3) please accept that PR so this PR can get the commit.

mvadu committed:
Fix current calculation

V1pr committed:
Merge pull request #3 from mvadu/fix_2130

Fix 2130 current setting and CS current scaling calculation

wolfmanjm commented:
That last commit was wrong IMO. The difference in current calculation is deliberate. The one suggested in the patch is RMS wheareas we have always used peak current. The .707 diiference is the difference between peak current and RMS current. Most motors seem to specify current in Peak, so that is what we have historically used. Changing it now is a mistake. At least it will not be changed for existing drivers as it would not be backward compatible. You really need to think carefully about the difference between Irms and Ipeak.

wolfmanjm commented:
If you want to set RMS current I suggest you use a different M code and config setting to do so. Or simply divide the setting you make by 0.707, that would then set the current to be RMS instead of Peak, but beware you don't burn out your motors.

V1pr commented:
> That last commit was wrong IMO. The difference in current calculation is deliberate. The one suggested in the patch is RMS wheareas we have always used peak current. The .707 diiference is the difference between peak current and RMS current. Most motors seem to specify current in Peak, so that is what we have historically used. Changing it now is a mistake. At least it will not be changed for existing drivers as it would not be backward compatible. You really need to think carefully about the difference between Irms and Ipeak.

@wolfmanjm I was not aware of this (never seen anywhere, my mistake probably), but in this case I'll revert it as  I totally agree with you, plus add a note everywhere, that the current is the PEAK not the RMS.

V1pr commented:
> If you want to set RMS current I suggest you use a different M code and config setting to do so. Or simply divide the setting you make by 0.707, that would then set the current to be RMS instead of Peak, but beware you don't burn out your motors.

I really don't think this is necessary, will just produce some extra code (and the PEAK setting also protects the novice users from burning the motors :D )

Commit:
reverted current calculations to use PEAK + added notes about it

f61 commented:
Good day! Where can I read a detailed configuration manual? I have a skr 1.3 tms 2209, kossel.

mvadu commented:
@wolfmanjm Hi Jim, the line `if (THEKERNEL->config->value(motor_driver_control_checksum, cs, sw_uart_rx_pin_checksum)->by_default("nc")->as_string() != "NC" )` in `src\modules\utils\motordrivercontrol\MotorDriverControl.cpp` is comparing default `nc` to `NC`, and looks like if I skip the config parameter the default is not used. Is that a typo not fixed or by design?

wolfmanjm commented:
looks likes a typo, it is not my code.

mvadu commented:
> looks likes a typo, it is not my code.

Ah, I missed to notice that section of code is being added as part of this PR, I will add a fix for this typo, as well to introduce the `printf` fix you added 2 days ago.

mvadu committed:
Fix typo in default nc compared with NC

mvadu committed:
Implement Jim's 8b5afad

Update streams->printf to printf

mvadu committed:
Check if RX pin is Interrupt Enabled

V1pr committed:
Merge pull request #4 from mvadu/fix_2209

Fixed NC typo + RX pin check, thanks mvadu!

superpirson commented:
Quick question, I was getting the message "TMC22x is running in READ ONLY mode on X [0], output status: 1" (This is from the M911 output [here](https://github.com/V1pr/Smoothieware/blob/53f3f57541fcdb7889438c545317f2fddf3d6504/src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp#L1576)  ) 
Shouldn't that say "WRITE ONLY" on the output instead of "READ ONLY"?

(apologies if I am totally out of my depth here)

mvadu commented:
> Quick question, I was getting the message "TMC22x is running in READ ONLY mode on X [0], output status: 1" (This is from the M911 output [here](https://github.com/V1pr/Smoothieware/blob/53f3f57541fcdb7889438c545317f2fddf3d6504/src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp#L1576) )
> Shouldn't that say "WRITE ONLY" on the output instead of "READ ONLY"?
> 
> (apologies if I am totally out of my depth here)

@superpirson you are right.. the message is wrong there.. @V1pr do you want to update it? I can send a PR if you want.

mvadu commented:
@wolfmanjm Jim, I have been testing this code for months now, and can confirm both SPI and UART drivers work reliably. I am not sure what your policy is on the PRs, but this looks like a feature complete to me.

mkotyk commented:
Hi All, I've been trying this branch out and keep seeing crashes.  My situation is a bit different, I'm trying to use this on a Bigtreetech SKR V1.4.  I've made a replacement soft-serial implementation that supports half duplex and non-interrupt pins.  Anyway, I'm just trying to figure out where this crash is.  I'm guessing 0xdeadbef5 is not a real address, and something wasn't initialized correctly.  I've connected with a SWD debugger and can hit breakpoints, but it's a bit like throwing darts.

Anyone know what that cryptic prompt at the bottom is?

```
TMC2209 using slave address 0 for axis X
Setting GCONF to 000001C5 for axis X
GCONF status: 33EFBEAD (871349933) [CRC: 1]
DRV status: 33EFBEAD (871349933) [CRC: 1]
MotorDriverControl INFO: configured motor X (1): as TMC2209, tx: 010A, rx: 010A

**Hard Fault**
  Status Register: 0x40000000
    Forced
**Bus Fault**
  Status Register: 0x82
    Fault Address: 0xdeadbef5
    Precise Data Access
$T0b0c:01010101;0d:187e0010;0e:f98d0400;0f:60820000;#e5
```

triffid commented:
That "cryptic prompt" is the mri debugging system trying to talk to GDB.

This sounds like exactly the sort of thing that happens when stack and heap
collide - which is highly likely when you start adding things to Smoothie
because it's already riding right on the limits of available memory.

On Tue, 10 Nov 2020 at 10:21, Mark Kotyk <notifications@github.com> wrote:

> Hi All, I've been trying this branch out and keep seeing crashes. My
> situation is a bit different, I'm trying to use this on a Bigtreetech SKR
> V1.4. I've made a replacement soft-serial implementation that supports half
> duplex and non-interrupt pins. Anyway, I'm just trying to figure out where
> this crash is. I'm guessing 0xdeadbef5 is not a real address, and something
> wasn't initialized correctly. I've connected with a SWD debugger and can
> hit breakpoints, but it's a bit like throwing darts.
>
> Anyone know what that cryptic prompt at the bottom is?
>
> TMC2209 using slave address 0 for axis X
> Setting GCONF to 000001C5 for axis X
> GCONF status: 33EFBEAD (871349933) [CRC: 1]
> DRV status: 33EFBEAD (871349933) [CRC: 1]
> MotorDriverControl INFO: configured motor X (1): as TMC2209, tx: 010A, rx: 010A
>
> **Hard Fault**
>   Status Register: 0x40000000
>     Forced
> **Bus Fault**
>   Status Register: 0x82
>     Fault Address: 0xdeadbef5
>     Precise Data Access
> $T0b0c:01010101;0d:187e0010;0e:f98d0400;0f:60820000;#e5
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1447#issuecomment-724406576>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABICGXWONIGQGUWZEJAOKLSPCPQXANCNFSM4LRMNYFA>
> .
>

mkotyk commented on ConfigSamples/Snippets/tmc2209_motor_driver.config:
```diff
@@ -0,0 +1,46 @@
+#motor_driver_control.beta.spreadcycle_tbl            32          # TBL    (16/24/32/48)
+#motor_driver_control.beta.spreadcycle_hstrt          4           # HSTART (1 - 8)
+#motor_driver_control.beta.spreadcycle_hend           0           # HEND   (-3 - 12)
+
```
Having some sane suggested values here would  be gerat.

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+    // TMC2209 specific config
+    if (chip_type == StepstickParameters::CHIP_TYPE::TMC2209) {
+        
+        uint8_t tcoolthrs   = THEKERNEL->config->value(motor_driver_control_checksum, cs, stallguard_tcoolthrs_checksum)->by_default(0)->as_int();
```
This value is a possible 20 bits, and will overflow a `uint8_t` - suggest using `int`

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+    }
+}
+
+uint8_t TMC220X::getStallguardResult(void)
```
Long being forced into `uint8_t` ... actual value is 10 bits according to datasheet.

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+    // Scaling to 10 bit is for compatibility to StallGuard2.
+    unsigned long result = readRegister(TMC220X_SG_RESULT_REGISTER);
+
+    return (uint8_t)result;
```
Dropping bit 8 on the floor.  Better to just return an int so the values make sense when  comparing to sgthreshold.

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+        serial(buf, 4, rbuf);
+        
+        // save response CRC
+        response_crc = rbuf[5];
```
Shouldn't the CRC be in `rbuf[7]`?

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+        // save response CRC
+        response_crc = rbuf[5];
+        
+        calc_crc(rbuf,4);
```
Shouldn't we be calculating crc on the whole message length of 8 bytes?  Will this put the calculated CRC in `rbuf[5]` and mess up the `i_datagram` below?

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+//reads a value from the TMC220X status register.
+unsigned long TMC220X::readRegister(int8_t reg_addr)
+{
+    uint32_t data;
+    uint8_t reg;
```
These variables aren't really needed.

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+
+bool TMC220X::isEnabled()
+{
+    if (chopconf_register_value & TMC220X_CHOPCONF_TOFF) {
+        return true;
+    } else {
+        return false;
+    }
```
```suggestion
      return (chopconf_register_value & TMC220X_CHOPCONF_TOFF);
```

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+
+bool TMC220X::isCurrentScalingHalfed()
+{
+    if (this->chopconf_register_value & TMC220X_CHOPCONF_VSENSE) {
+        return true;
+    } else {
+        return false;
+    }
```
```suggestion
    return (this->chopconf_register_value & TMC220X_CHOPCONF_VSENSE);
```

mkotyk commented on src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp:
```diff
@@ -0,0 +1,1870 @@
+    }
+
+    //save the threshold value
+    this->sgthrs_register_value = threshold;
```
Not actually sending this value ever!!!

vonvogel007 commented:
Hello dear smoothie professionals!
I wish you a happy new year!
I upgraded my K40 laser with a SKR V1.3 with TMC2208 and it works quite well. The problem is that the 2208 cannot be set via UART. I put the jumpers for UART on the SKR V1.3 and the lines are available in the config.txt
"uart0.baud_rate 115200"
"second_usb_serial_enable false"
For testing I set the drive currents to 0.1 in the config, but the drive currents that are set via the potentiometer are used!
I've been looking for a solution for a few days, but can't find the problem. Is there a special firmware that I have to use?

wolfmanjm commented:
Sadly we neither support skr boards nor do we support TMC2208. I suggest you contact SKR for help.
Or contact the author of this PR to see if he can help you.
This PR is not part of the official Smoothie build and in unlikely to be included.

vonvogel007 commented:
Thank you for the fast answer!
It can't be the board because with Marlin it works with UART

arthurwolf commented:
This isn't about the hardware of the board itself necessarily. Maybe this
would help get some context:
http://smoothieware.org/troubleshooting#what-is-wrong-with-mks

On Sat, Jan 9, 2021 at 7:50 PM vonvogel007 <notifications@github.com> wrote:

> Thank you for the fast answer!
> It can't be the board because with Marlin it works with UART
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1447#issuecomment-757350480>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA2SFJ5PZLUWXOX53V3Y6LSZCQN5ANCNFSM4LRMNYFA>
> .
>


-- 

勇気とユーモア

V1pr commented:
> Sadly we neither support skr boards nor do we support TMC2208. I suggest you contact SKR for help.
> Or contact the author of this PR to see if he can help you.
> This PR is not part of the official Smoothie build and in unlikely to be included.

"Highly unlikely"? :( I'm sad :( Makes me think, that my work was a bit pointless.

V1pr commented:
> This isn't about the hardware of the board itself necessarily. Maybe this would help get some context: http://smoothieware.org/troubleshooting#what-is-wrong-with-mks
> […](#)
> On Sat, Jan 9, 2021 at 7:50 PM vonvogel007 ***@***.***> wrote: Thank you for the fast answer! It can't be the board because with Marlin it works with UART — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#1447 (comment)](https://github.com/Smoothieware/Smoothieware/pull/1447#issuecomment-757350480)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAA2SFJ5PZLUWXOX53V3Y6LSZCQN5ANCNFSM4LRMNYFA> .
> -- 勇気とユーモア

@arthurwolf however I agree with you on MKS, but SKR is open-source... there is everything on their public github (schematics etc)

V1pr commented:
> Thank you for the fast answer!
> It can't be the board because with Marlin it works with UART

@vonvogel007 check your wirining an PIN assignment. Marlin uses a lot of thing an other way (and read the posts from this thread, someone already bumped into this).

wolfmanjm commented:
@V1pr it is not a waste of your time, even though it may never get into the main build of smoothie (because it is really only useful to non smoothieboard users, and I have no way to test it), it is useful to those that are using these driver chips.
Eventually once this builds cleanly against the current build, it would be nice to put this in the third party branches so those that need it can easily access it.
One of the issues I have with merging this is that going forward I would have to maintain it whenever I make any changes to the rest of the code, and not being able to test it would be a problem for me.

arthurwolf commented:
> @arthurwolf however I agree with you on MKS, but SKR is open-source... there is everything on their public github (schematics etc)

Last time I looked, it was the most classic/obvious "pretend open-source", showing only schematics and no source files, and not providing a proper open-source license. It's possible they fixed that since, but that would surprise me *a lot*.

> "Highly unlikely"? :( I'm sad :( Makes me think, that my work was a bit pointless.

A lot of work isn't integrated in the main branch for many reasons, mostly because we don't feel like we can adequately support it in the long run. That's what we have smoothieware.org/third-party-branches for, and many branches are listed there and are in fact used by many, so any work that ends up going there instead of the main branch is definitely not pointless.

arthurwolf commented:
@V1pr Yep, just looked, it's *definitely* not Open-Source. Having a github account where you post some vague documentation of the hardware doesn't make any of this Open-Source. Open-Source would "amongst many other things* require that I be able to modify it ( that would require the *source* which they don't provide ), and allowed to modify it (through a proper Open-Source license) which they don't allow either.

This is about as Open-Source as Adobe Photoshop.

vonvogel007 commented:
If you take opensource literally then it is the disclosure of sources, construction plans, program code, etc. The circuit diagrams from the SKR are published and so everyone can rebuild or change the circuit board! For me this is open source!
Unfortunately, I have no idea about C as a programming language and have always avoided dealing with it. I come more from Borland Pascal, Delphi, Purebasic. As a controller for our school projects (I am a teacher at a school) I have been using PICAXE for many years. Of course, the people who put on an open source decide for themselves what they do or not because they do it for free, I have absolute understanding! On the other hand, a lot of energy is lost with opensource because several developers develop in different directions and then work done as by V1pr may not be further disseminated. This is frustrating for these people! Or there are spin-offs like Linux where there are now hundreds of different Linux versions. That is uneconomical and pointless because it only leads to the system never developing as a whole as it actually could. But I digress!
Why can't you just insert a "Type of Stepperdriver" parameter in the config.txt which then controls how the driver is controlled in the firmware? Then the work of V1pr could be meaningfully integrated. The more universal the open source project, the better known it is and the greater the recognition!

arthurwolf commented:
> For me this is open source!

And for some people China is a Democracy.

It's pretty crazy you can think their stuff is Open-Source, when they *do not provide the SOURCE* files of their work. Really? There's literally no source files, and it's Open-Source. That's an Olympic stretch...

 I don't care much about vague opinions, I care about what's clearly established. 

Open-Source is very clearly defined:
* https://opensource.org/osd
* https://en.wikipedia.org/wiki/Open_source
* https://www.oshwa.org/definition/
* https://en.wikipedia.org/wiki/Open-source_hardware

Smoothie is Open-Source. SKR clearly isn't (despite having no chance of ever existing without the work of the thousands of Open-Source workers that came before them).

Publishing PDFs of the circuit or schematic is simply not Open-Source, at all. You're also ignoring the fact that if you take those diagrams and modify them (ignoring the huge inconvenience of not having the source), you would be BREAKING THE LAW by doing so.

Can you really be so obtuse you think that them *not allowing you to modify their work* is Open-Source. I'm sorry, you must be *very very* new to the notion of Open-Source if this makes sense to you.

I'm not going to comment much on the stuff about Linux, thousands of versions, energy lost, etc, that's your opinion, it's extremely easily demonstrated wrong, but I have zero interest in having that conversation. It does make it very obvious you have only very superficial understanding of Open-Source and how it works, which does make a lot of sense considering the "to me this is Open-Source" position. This is an Open-Source project though, so don't be surprised if WE take it seriously and know a lot about it.

> Why can't you just insert a "Type of Stepperdriver" parameter in the config.txt which then controls how the driver is controlled in the firmware? 

That's exactly how things work right now. That would still required including his code, and having to maintain it, which is problematic for many reasons, which have been explained before, and you seem not to make a lot of efforts to understand.

arthurwolf commented:
> Can you really be so obtuse you think that them not allowing you to modify their work is Open-Source. 

@vonvogel007 Through a separate and completely unrelated conversation, I've been made aware that the use of the word "obtuse" in English could be taken as insulting (meaning "stupid" or something like that). 

I'm natively a French speaker, obtuse to me and in this context is not about the person but the argument, and it doesn't mean stupid but rather lacking perspective/open-mindedness (in the same way an obtuse angle is "narrow" and not as open, the idea transposes to arguments not taking full context into account and/or mistakenly lacking some elements that would help see things otherwise).

So I didn't in any way mean to be insulting here, I was making a comment on the argument itself, and I apologize if you felt in any way insulted by the comment.

I'm still curious/interested to know what you think about this, I feel like the case for why calling SKR Open-Source isn't justified/is an extreme stretch, is a very solid one, and if at this point you still are not convinced of this, I really would like to know why. I understand if you're not going to keep that conversation going, just expressing what I'd like to happen/the fact I'm curious about it.

vonvogel007 commented:
Hi Arthur,

can you tell me again what it was about?
I can't find the post right now!
I don't think Github is that great as a forum!

Wolfgang

Am 15.02.2021 16:06, schrieb Arthur Wolf:
>
>     Can you really be so obtuse you think that them not allowing you
>     to modify their work is Open-Source.
>
> @vonvogel007 <https://github.com/vonvogel007> Through a separate and 
> completely unrelated conversation, I've been made aware that the use 
> of the word "obtuse" in English could be taken as insulting (meaning 
> "stupid" or something like that).
>
> I'm natively a French speaker, obtuse to me is not about the person 
> but the argument, and it doesn't mean stupid but rather lacking 
> perspective/open-mindedness (in the same way an obtuse angle is 
> "narrow" and not as open, the idea transposes to arguments not taking 
> full context into account and/or mistakenly lacking some elements that 
> would help see things otherwise).
>
> So I didn't in any way mean to be insulting here, I was making a 
> comment on the argument itself, and I apologize if you felt in any way 
> insulted by the comment.
>
> I'm still curious/interested to know what you think about this, I feel 
> like the case for why calling SKR Open-Source isn't justified/is an 
> extreme stretch, is a very solid one, and if at this point you still 
> are not convinced of this, I really would like to know why. I 
> understand if you're not going to keep that conversation going, just 
> expressing what I'd like to happen/the fact I'm curious about it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Smoothieware/Smoothieware/pull/1447#issuecomment-779280123>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AP6HPSQV6CRLF2L7C2N54O3S7EZ7ZANCNFSM4LRMNYFA>.
>

arthurwolf commented:
@vonvogel007 It's not a forum at all :) Are you receiving this via email? If so the pull request this is about should be at the bottom of the email. Just in case, here's the link: https://github.com/Smoothieware/Smoothieware/pull/1447#issuecomment-779371742

vonvogel007 commented:
Hello Artur,
just found the page again.
If I see it correctly then it is about that you want to explain to me what is open source and what is not !?
You are right if you assume that I don't want to discuss this any further! My time is too valuable for that and I prefer to deal with solving technical problems!
And ... since I am a young beginner, as you have noticed, you can't learn anything from me anyway and I am probably too old to learn anything from you! ;-)

arthurwolf commented:
@vonvogel007 Last month we had a conversation about what Open-Source means, in which both of us presented arguments. After a while and both of us presented what we thought, on the last post before you stopped answering, I used the word "obtuse". Since then (a few days back), I learned that to native speakers, obtuse is insulting. So I came back here to make sure to make it clear I didn't mean to use the word in an insulting fashion. That's it. Just making sure you don't think I'm insulting you when I'm not.

About Open-Source, if you don't want to get to the bottom of the disagreement, I can certainly not force you to discuss it, so if you choose the conversation is over, it is.

Cheers.
